### ISSUE TITLE
DataBake: split serialized form from runtime form

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#2452]
A general DataBake feature that would solve several problems, including #2417, would be a way to specify a different data struct as the serialized form of the main data struct, with functions to convert between them.

For example, you might write

```rust
#[databake(path = foo)]
#[databake::serialize_as(path = foo::FooFields, from = from_fields, to = to_fields)]
pub struct Foo<'data> {
  // private fields
}

pub struct FooFields {
  // public fields
}

impl Foo<'data> {
  pub const fn from_fields(fields: &'data FooFields) -> Self {
    // a cheap const constructor
  }
  pub fn to_fields(&self) -> FooFields {
    // invoked during serialization only
  }
}
```

CC @Manishearth @robertbastian

**Comment by Manishearth** [comment]
The way I'd prefer to do this is:

```rust
#[databake(path = foo)]
#[databake::constructor = from_fields]
pub struct Foo<'data> {
  // private fields (NOT processed by the proc macro)
}

impl<'data> Foo<'data> {
  // fields are in the same order as they are in Foo.
  // if order needs to change you must switch to fully manual
  pub const fn from_fields(field1: Field1Type, field2: Field2Type ) -> Self {
  }
}
```

We could also have both (serde does), but I consider having methods to be less API-polluting than having a new struct. Though in both cases we could `doc(hidden)` them (I'm more open to doc-hidden on methods and structs over fields)

**Comment by sffc** [comment]
@Manishearth Your proposal is tied to the list of fields in `Foo`. This serves a more narrow use case. My proposal decouples the list of fields in `Foo` from the struct being serialized, which fixes the private fields issue but also solves other issues.

Also, you copied my comment

```
  // private fields (NOT processed by the proc macro)
```

which is exactly what your solution _doesn't_ do.

**Comment by Manishearth** [comment]
oops, yeah.

Your proposal does decouple things completely, but at the cost of a new public type; which I consider to be pretty high. I would rather ICU4X not be peppered with these fields types.

**Comment by sffc** [comment]
Pros and cons of a separate fields type:

1. Pro: Allows the type `Foo` to have a different data model / different private fields
2. Pro: The data model of `Foo` can change without changing any public function signatures
3. Pro: We could additionally consider `FooFields` as something that could evolve if we say that it is unstable (like the unstable constructors)
4. Con: ICU4X peppered with more public internal types
5. Con: A bit more boilerplate than Manish's proposal (but still less than custom Bake impls)

Did I miss anything?

**Comment by sffc** [comment]
Another thought on the proposed convention for function names: `from_baked_fields` and `to_baked_fields`. These function names can be the default names so that they don't need to be listed in the derive. The standard boilerplate then becomes

```rust
#[databake(path = foo)]
#[databake::serialize_as(foo::FooBakedFields)]
pub struct Foo<'data> {
  // private fields (NOT processed by the proc macro)
}

#[databake(path = foo)]
pub struct FooBakedFields {
  // public fields (processed by the proc macro)
}

impl<'data> Foo<'data> {
  pub const fn from_baked_fields(fields: &'data FooBakedFields) -> Self {
    // a cheap const constructor
  }
  pub fn to_baked_fields(&self) -> FooBakedFields {
    // invoked during serialization only
  }
}
```

**Comment by sffc** [comment]
Thought: This would also enable us to perform validation (to the extent allowed in a const context) either by panicking in `from_baked_fields` or by returning a `Result` and making DataBake generate a panicky match statement.

**Comment by Manishearth** [comment]
I mean, part of the thing is that we can have _both_ these schemes, like serde does. I just don't want us to use from_fields in ICU4X. Perhaps with a hidden type, I guess.

The problem with having a proposed convention is that that's super rigid since this is an impact on the public API. Unlike serde where everything can be module-private; any such function must be public; so I'd argue that we _have_ to let the user choose the names. That's also why I want more flexibility in not requiring the user to use e.g. FooBakedFields or whatever. Everything of this form has an impact on the public API and that's quite heavy for what is essentially a plumbing function for a library.

**Comment by sffc** [comment]
I'm not proposing that any of the names are fixed; they should be customizable. I'm only proposing that `from_baked_fields` and `to_baked_fields` seems like a reasonable convention to be used in absence of something better.

The functions and struct can be `#[doc(hidden)]` if desired.

**Comment by Manishearth** [comment]
Hmm, fair. I think that would be okay for ICU4X.

**Comment by sffc** [comment]
CC @robertbastian PTAL

**Comment by robertbastian** [comment]
From the discussion so far I think this is hard to get right, and I don't really see a benefit. If structs have private fields they will have a custom serde implementations, so I don't think avoiding the custom `Bake` implementation is that important. 

Validation is nice to have, but as long as we cannot validate `ZeroVec`s, this would not solve 99% of the unsafe calls in the generated code, and we just have to trust that it's correct.

**Comment by sffc** [comment]
Discussion: @robertbastian: for `Char16Trie`, for example, we can do

```
#[bake(constructor = from_fields, args = data)]
pub struct Char16Trie<'data> {
    /// An array of u16 containing the trie data.
    #[cfg_attr(feature = "serde", serde(borrow))]
    data: ZeroVec<'data, u16>,
}

impl Char16Trie {
    pub fn from_fields(data: &'data [u16]) { ... }
}
```

**Comment by sffc** [comment]
Discuss with:

- @robertbastian 
- @sffc 

Optional:

- @Manishearth

**Comment by sffc** [comment]
A little research and discussion with @robertbastian 

What I'm asking for here is basically a different derive that uses an encoding method and a decoding method instead of a struct construction impl.

The encoded type does not need to be a custom type. It could be just bytes.

This would impact the following custom impls in the repo:

- `impl databake::Bake for LstmDataFloat32<'_>`
- `impl databake::Bake for PluralElementsPackedULE<V: VarULE + ?Sized>` (maybe)
- `impl databake::Bake for UnitsDisplayName<'_>`
- `impl databake::Bake for CodePointInversionList<'_>`
- `impl databake::Bake for CurrencyExtendedData<'_>`

These impls are all ones that use various types of `from_parts_unchecked` constructors.

Conclusion: could be done, but @robertbastian thinks it is more gnarly to have the custom derive than to have the five impls. @sffc thinks that this is a pattern we want to use more, but acknowledges that the status quo isn't _too bad_.

**Comment by Manishearth** [comment]
> These impls are all ones that use various types of `from_parts_unchecked` constructors.

Wouldn't the ctor method also need from_parts_unchecked constructors? I'm a bit lost as to the actual benefit here.

**Comment by robertbastian** [comment]
Yes, this would not change the baked output, it would just automate the custom impls.

**Comment by sffc** [comment]
My implicit axiom is generally that the derive does two things: it reduces boilerplate, and it encourages a canonical way of doing things. I see the ability to decouple the "serialized" bake struct from the runtime bake struct to be fundamental and a pattern we want to encourage.

The five manual impls that we could automate solve a variety of issues:

- `LstmDataFloat32` and `CodePointInversionList` have manual impls because of invariants between fields
- `UnitsDisplayName` and `CurrencyExtendedData` have manual impls to improve compile times (#5230)
- `PluralElementsPackedULE` has a manual impl I think because of complexities involving the generic `V`

**Comment by Manishearth** [comment]
I'm still lost as to the benefit. Either way we need to write some custom code doing the conversion. Can you give concrete examples of what you see of as a before and after?

I don't really see this concept of a "serialize" vs "runtime" bake struct as fundamental; I think it's one particular pattern amongst many, but perhaps I'm misunderstanding.

(My contention with the code example in the original issue is precisely that it seems to privilege one particular pattern for doing this type of thing, a pattern that I don't see as inherently better)

**Comment by Manishearth** [comment]
A different question: what is the cost you perceive of the current code, that you are trying to ameliorate?

**Comment by sffc** [comment]
Before (current):

```rust
/// Currency Extended  data struct.
#[derive(Debug, Clone, PartialEq, yoke::Yokeable, zerofrom::ZeroFrom)]
#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
#[yoke(prove_covariance_manually)]
pub struct CurrencyExtendedData<'data> {
    #[cfg_attr(feature = "serde", serde(borrow))]
    pub display_names: PluralElementsPackedCow<'data, str>,
}

icu_provider::data_struct!(CurrencyExtendedData<'_>, #[cfg(feature = "datagen")]);

impl<'data> CurrencyExtendedData<'data> {
    /// Construct an instance directly from a byte slice.
    ///
    /// # Safety
    ///
    /// The bytes must represent a valid [`icu_plurals::provider::PluralElementsPackedULE`]
    pub const unsafe fn from_bytes_unchecked(bytes: &'data [u8]) -> Self {
        Self {
            display_names: icu_plurals::provider::PluralElementsPackedCow {
                elements: alloc::borrow::Cow::Borrowed(
                    // Safety: this function's safety invariant guarantees that the bytes
                    // represent a valid `PluralElementsPackedULE`
                    icu_plurals::provider::PluralElementsPackedULE::from_bytes_unchecked(bytes),
                ),
            },
        }
    }
}

#[cfg(feature = "datagen")]
impl databake::Bake for CurrencyExtendedData<'_> {
    fn bake(&self, ctx: &databake::CrateEnv) -> databake::TokenStream {
        use zerovec::ule::VarULE;
        ctx.insert("icu_experimental::dimension::provider::extended_currency");
        let bytes = self.display_names.elements.as_bytes().bake(ctx);
        // Safety: The bytes are returned by `PluralElementsPackedULE::slice_as_bytes`.
        databake::quote! { unsafe {
            icu_experimental::dimension::provider::extended_currency::CurrencyExtendedData::from_bytes_unchecked(#bytes)
        }}
    }
}
```

After:

```rust
/// Currency Extended  data struct.
#[derive(Debug, Clone, PartialEq, yoke::Yokeable, zerofrom::ZeroFrom)]
#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
#[cfg_attr(feature = "databake", derive(databake::Bake))]
#[cfg_attr(feature = "databake", databake(path = zerotrie, custom_repr))]  // <-- NEW "custom_repr"
#[yoke(prove_covariance_manually)]
pub struct CurrencyExtendedData<'data> {
    #[cfg_attr(feature = "serde", serde(borrow))]
    pub display_names: PluralElementsPackedCow<'data, str>,
}

icu_provider::data_struct!(CurrencyExtendedData<'_>, #[cfg(feature = "datagen")]);

impl<'data> CurrencyExtendedData<'data> {
    /// Construct an instance directly from a byte slice.
    ///
    /// # Safety
    ///
    /// The bytes must be returned by `to_databake_repr`, which are the bytes
    /// of a valid `PluralElementsPackedULE`
    pub const unsafe fn from_databake_repr(bytes: &'data [u8]) -> Self {
        Self {
            display_names: icu_plurals::provider::PluralElementsPackedCow {
                elements: alloc::borrow::Cow::Borrowed(
                    // Safety: this function's safety invariant guarantees that the bytes
                    // represent a valid `PluralElementsPackedULE`
                    icu_plurals::provider::PluralElementsPackedULE::from_bytes_unchecked(bytes),
                ),
            },
        }
    }

    /// Returns the bytes of the inner `PluralElementsPackedULE`
    pub fn to_databake_repr(&self) -> &[u8] {
        self.display_names.elements.as_bytes()
    }
}
```

**Comment by sffc** [comment]
I'm not sure where my messaging is getting lost. I think I've been pretty clear, and I've even tried to decode my personally held axioms. But I'll try again:

Whereas, derives play a major role in nudging clients toward certain design patterns; and

Whereas, with the current state of affairs, the derive forces clients to couple their serialized data and runtime struct; and

Whereas, decoupling serialized data from the runtime struct allows clients of databake to reduce code size, reduce compile times, and uphold invariants of private fields, as evidenced by the ICU4X call sites; and

Whereas, other libraries such as Serde allow this behavior in their derive via `#[serde(try_from)]`, demonstrating the utility for the feature in the broader ecosystem, not just ICU4X; be it

Resolved, that the databake derive should allow decoupling runtime and serialized forms.

---

Put another way: imagine you are mentoring a new or infrequent ICU4X contributor, and they are introducing a data struct that needs the decoupling in order to meet code size or compile time goals. You want this process to be, "add this line to your databake derive, and implement these two functions with well-defined behavior, as you can see in the nice docs page", not "remove the databake derive and go implement it on your own, using this other code over here as an example".

**Comment by Manishearth** [comment]
Okay, I understand now. I misunderstood the problem being solved, first I thought the problem being solved was the proliferation of unchecked (which I noted, and Robert said it doesn't change anything), then I thought the problem being solved was that the "fundamental" pattern of "serialized" vs "runtime" types was not being used enough, which does not sound intrinsically like a problem to me (it is a solution). Understanding the core problem 


Distracted by this, I didn't realize that "The encoded type does not need to be a custom type. It could be just bytes." solved my original concern of "I don't want a blessed databake pattern to force you to make new public types" (or, if we do end up doing that, we should make it subtler so that it gets auto-hidden or something).


-----


Stating my current interpretation concretely: the underlying problem being solved appears to be that sometimes you want to do non-fields-based construction. _Often_ it is from a single "serialized" type. We wish to have patterns that nudge people away from having to write a boilerplatey `DataBake` impl.

I agree with the problem as stated and think it's worth solving.


I agree less with the fundamentality of the "serialized" type concept; I think there are other forms of non-fields-based construction (e,g, from 2+ parameters). _However_ I think they will be more rare, and the "serialized" model has a lot of nice affordances when it comes to designing a solution: anything involving more than one intermediate type and suddenly you need to tell the derive much more information about what's going on.

---------

Looking at the code as presented, I think this is a good solution for the problem at hand. We may wish to allow users to customize the `custom_repr` functions with `custom_repr(from_name, to_name)` but that's an optional enhancement. In ICU4X I would like to encourage users to use existing types as intermediates rather than making new ones.

**Comment by sffc** [comment]
Yes, the part about "the serialized form could just be bytes" is an evolution from when I originally made this issue in 2022, and I think in most cases we will be able to do this without adding an extra type.

**Comment by robertbastian** [comment]
Some more discussion with @sffc: 

```rust

#[derive(Bake)]
#[bake(destructor = "Self::as_bytes", constructor = "Self::from_bytes")]
struct Foo;

// generates
impl Bake for Foo {
    fn bake(&self, env: ...) -> TokenStream {
        let a = Self::as_bytes(&self);
        quote! {
            Foo::from_bytes(#a)
        }
    }
}

#[derive(Bake)]
#[bake(destructor = "Self::as_bytes", constructor = "Self::from_bytes", safety_comment = "from_bytes is safe for as_bytes")]
struct Foo;

// generates
impl Bake for Foo {
    fn bake(&self, env: ...) -> TokenStream {
        let a = Self::as_bytes(&self);
        quote! {
            // Safety: from_bytes is safe for as_bytes
            unsafe { Foo::from_bytes(#a) }
        }
    }
}
```

**Comment by Manishearth** [comment]
(I have complicated thoughts on this from a safety and API perspective, please wait till I return to merge)

**Comment by sffc** [comment]
My current PR is in #6576; the code in @robertbastian's post above is a strawperson about how this could be done without a trait. This isn't urgent and I won't land anything until you've had the opportunity to review.

**Comment by Manishearth** [comment]
Yep, I'm aware, I saw the PR go by and also had complex thoughts about it.

**Comment by Manishearth** [comment]
(posted thoughts on PR, let's continue discussion there)
